crates/index/src/optimizing/indexing.rs

```diff
@@ -99,26 +99,29 @@ impl<O: Op> OptimizerIndexing<O> {
             }),
         )
     }
-    fn main(self, shutdown: Receiver<Infallible>) {
+    fn main(self, shutdown_rx: Receiver<Infallible>) {
         let index = self.index;
         loop {
             let view = index.view();
             let threads = view.flexible.optimizing_threads;
+            let (finish_tx, finish_rx) = bounded::<Infallible>(1);
             rayon::ThreadPoolBuilder::new()
                 .num_threads(threads as usize)
                 .build_scoped(|pool| {
                     std::thread::scope(|scope| {
-                        scope.spawn(|| match shutdown.recv() {
-                            Ok(never) => match never {},
-                            Err(RecvError) => {
-                                pool.stop();
-                            }
+                        let handler = scope.spawn(|| {
+                            monitor(&finish_rx, &shutdown_rx);
+                            pool.stop();
```

Comment:
It should be called only if the reason why `monitor` returned is `shutdown_rx` disconnected.